packages/@lwc/engine-core/src/framework/freeze-template.ts

```diff
@@ -0,0 +1,130 @@
+    'sort',
+    'fill',
+    'splice',
+    'copyWithin',
+] as const;
+
+function getOriginalArrayMethod(prop: typeof ARRAY_MUTATION_METHODS[number]) {
+    switch (prop) {
+        case 'pop':
+            return ArrayPop;
+        case 'push':
+            return ArrayPush;
+        case 'shift':
+            return ArrayShift;
+        case 'unshift':
+            return ArrayUnshift;
+        case 'reverse':
+            return ArrayReverse;
+        case 'sort':
+            return ArraySort;
+        case 'fill':
+            return ArrayFill;
+        case 'splice':
+            return ArraySplice;
+        case 'copyWithin':
+            return ArrayCopyWithin;
+    }
+}
+
+let mutationWarningsSilenced = false;
+
+// Warn if the user tries to mutate tmpl.stylesheets, e.g.:
+// `tmpl.stylesheets.push(someStylesheetFunction)`
+function warnOnArrayMutation(stylesheets: TemplateStylesheetFactories) {
+    // We can't handle users calling Array.prototype.slice.call(tmpl.stylesheets), but
+    // we can at least warn when they use the most common mutation methods.
+    for (const prop of ARRAY_MUTATION_METHODS) {
+        const originalArrayMethod = getOriginalArrayMethod(prop);
+        stylesheets[prop] = function arrayMutationWarningWrapper() {
+            logError(
+                `Mutating the "stylesheets" array on a template function ` +
+                    `is deprecated and may be removed in a future version of LWC.`
+            );
+            // @ts-ignore
+            return originalArrayMethod.apply(this, arguments);
+        };
+    }
+}
+
+// TODO [#2782]: eventually freezeTemplate() will _actually_ freeze the tmpl object. Today it
+// just warns on mutation.
+export function freezeTemplate(tmpl: Template) {
+    if (process.env.NODE_ENV !== 'production') {
+        if (!isUndefined(tmpl.stylesheets)) {
+            warnOnArrayMutation(tmpl.stylesheets);
+        }
+        for (const prop of TEMPLATE_PROPS) {
+            let value = tmpl[prop];
+            defineProperty(tmpl, prop, {
+                enumerable: true,
+                configurable: true,
+                get() {
+                    return value;
+                },
+                set(newValue) {
+                    if (!mutationWarningsSilenced) {
+                        logError(
+                            `Dynamically setting the "${prop}" property on a template function ` +
+                                `is deprecated and may be removed in a future version of LWC.`
+                        );
+                    }
+                    value = newValue;
+                },
+            });
+        }
+
+        const originalDescriptor = getOwnPropertyDescriptor(tmpl, 'stylesheetTokens');
+        defineProperty(tmpl, 'stylesheetTokens', {
+            enumerable: true,
+            configurable: true,
+            get: originalDescriptor!.get,
+            set(value) {
+                if (process.env.NODE_ENV !== 'production') {
```

Comment:
This whole of `freezeTemplate` is already wrapped in `if (process.env.NODE_ENV !== 'production') {`, so this check is redundant.

Reply:
Fixed.